crates/ruff_formatter/src/builders.rs

```diff
@@ -359,31 +358,62 @@ impl std::fmt::Debug for DynamicText<'_> {
     }
 }
 
-/// Creates a text from a dynamic string and a range of the input source
-pub fn static_text_slice(text: Rc<str>, range: TextRange) -> StaticTextSlice {
-    debug_assert_no_newlines(&text[range]);
+/// Emits a text as it is written in the source document. Optimized to avoid allocations.
+pub const fn source_text_slice(
+    range: TextRange,
+    new_lines: ContainsNewLines,
+) -> SourceTextSliceBuilder {
+    SourceTextSliceBuilder { range, new_lines }
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum ContainsNewLines {
+    /// The string contains new line characters
+    Yes,
+    /// The string contains no new line characters
+    No,
```

Comment:
These are mainly optimizations to avoid searching the string for newlines in release builds. It makes the API a bit more cumbersome to use but it's good for performance because we almost always know whether a token can contain new lines